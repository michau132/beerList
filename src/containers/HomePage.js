import React, { Component } from 'react';
import { observer, inject, PropTypes as MobXPropTypes } from 'mobx-react';
import styled from 'styled-components';
import BeerList from '../components/BeerList';
import Header from '../components/Header';


const Main = styled.main`
  overflow-y: auto;
`;

@inject('BeerListStore')
@observer
class HomePage extends Component {
  static propTypes ={
    BeerListStore: MobXPropTypes.observableObject.isRequired,
  }

  constructor(props) {
    super(props);
    this.scroller = React.createRef();
    this.state = {
      isCorrectUrl: true,
    };
  }


  componentDidMount() {
    const { BeerListStore, match: { params: { id } } } = this.props;
    // eslint-disable-next-line no-restricted-globals
    if (isNaN(id)) {
      this.setState({
        isCorrectUrl: false,
      });
    } else {
      BeerListStore.getBeers();
    }
  }

  handleScroll = () => {
    const { BeerListStore } = this.props;
    const { loading, error, lastBeer } = BeerListStore;
    const { scroller } = this;
    if (
      scroller
      && (scroller.scrollHeight - scroller.scrollTop) < 1000
      && !loading
      && !error
      && !lastBeer
    ) {
      BeerListStore.getBeers();
    }
  }


  render() {
    const { isCorrectUrl } = this.state;
    return (
      <Main
        onScroll={this.handleScroll}
        ref={(scroller) => {
          this.scroller = scroller;
        }}
      >
        <Header />
        <BeerList />
        {!isCorrectUrl && (
        <p>
          Cannot get data, provided url at the end does not match number.
          Correct url should replace
          {' '}
          <strong>{`${window.location.hash.substring(2)} `}</strong>
           with number like
          {' '}
          <strong>1</strong>
          .
        </p>
        )}
      </Main>
    );
  }
}

export default HomePage;
