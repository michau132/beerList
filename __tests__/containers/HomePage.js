import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { observable } from 'mobx';
import HomePage from '../../src/containers/HomePage';

const BeerListStore = observable({
  getBeers: jest.fn(),
  loading: false,
  error: null,
  lastBeer: false,
});

const props = {
  BeerListStore,
  match: { params: { id: 2 } },
  location: { pathname: '/2' },
};

describe('testing HomePage with opened Modal', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<HomePage.wrappedComponent {...props} />);
  });
  test('renders without crashing', () => {
    expect(wrapper).toBeDefined();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should fire ComponentDidMount', () => {
    wrapper.instance().componentDidMount();
    expect(props.BeerListStore.getBeers).toHaveBeenCalled();
  });

  test('should call onScroll fn', () => {
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'handleScroll');
    instance.forceUpdate();
    wrapper.prop('onScroll')();
    expect(spy).toHaveBeenCalled();
  });
});
