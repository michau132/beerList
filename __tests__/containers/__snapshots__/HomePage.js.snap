// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing HomePage with opened Modal renders without crashing 1`] = `
<styled.main
  onScroll={[Function]}
>
  <Header />
  <inject-_class-with-BeerListStore />
</styled.main>
`;
